### PR TITLE
Security Fix for Regular Expression Denial of Service (ReDoS) - huntr.dev

### PR DESCRIPTION
**Pull request by huntr-helper**
https://huntr.dev/users/alromh87 has fixed the Regular Expression Denial of Service (ReDoS) vulnerability 🔨. alromh87 has been awarded $25 for fixing the vulnerability through the huntr bug bounty program 💵. Think you could fix a vulnerability like this?

Get involved at https://huntr.dev/

Q | A
Version Affected | ALL
Bug Fix | YES
Original Pull Request | https://github.com/418sec/locutus/pull/3
Vulnerability README | https://github.com/418sec/huntr/blob/master/bounties/npm/locutus/1/README.md

### ⚙️ Description *

locutus is vulnerable to ReDoS. The regular expression at src/php/network/inet_pton.js:24 is vulnerable to ReDoS. It is possible to cause increasing slow-downs which lock the event loop by passing strings which have some number of repeating a characters followed by a . character. For example: aaaaaaaaaaaa.

###   Technical Description *

Fixed by replacing vulnerable reg exp with algorithm using split and verifying chunks

Updated with latest master from upstream

###   Proof of Concept (PoC) *

1. Create the following PoC file:

```js
// poc.js
const inet_pton = require('locutus/php/network/inet_pton');
console.log('starting')
for (let i=4; i < 40; i++)
{
    const str = new Array(i).fill('a').join('') + '.';
    const start = + new Date();
    inet_pton(str);
    const elapsed = (+ new Date()) - start;
   
    console.log('i=' + i + ' elapsed (seconds): ' + elapsed / 1000);
}
```

2. Execute the following commands in another terminal:

```bash
npm i locutus # Install affected module
node poc.js #  Run the PoC
```

3. Check the Output: elapsed time increments drastically

loctusPOC

###   Proof of Fix (PoF) *

After fix the amount of time needed to identify invalid ip is minimized:

loctusPOF

###   User Acceptance Testing (UAT)

Library decodes IPv6 normally can be tested with:

```
const inet_pton = require('./src/php/network/inet_pton.js');

console.log('\n-------------------\nValid:')
printResult(inet_pton('::'));
printResult(inet_pton('::1'));
printResult(inet_pton('1::1'));
printResult(inet_pton('2001:0db8:0000:0000:0000:ff00:0042:8329'));
printResult(inet_pton('127.0.0.1'));

function printResult(result){
  if(typeof result !== 'string')return;
  for (let j=0; j<result.length; j++){
    process.stdout.write(`${result.charCodeAt(j)}-`);
  }
  console.log('');
}
```

loctusOK

loctusOrig

Remaining Regexp is safe for ReDOS
Captura de pantalla de 2020-10-21 18-56-52

### CODE REVIEW

**Comment by JamieSlome** [comment]
@kukawski - let me know if you have any thoughts or questions!

Cheers! 🍰

**Comment by kukawski** [comment]
@JamieSlome thanks for the PR with the fix. I'll review it asap.

**Comment by JamieSlome** [comment]
@kukawski - Great!

File: src/php/network/inet_pton.js [hunk 1]
```diff
@@ -21,36 +21,44 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
+
+  let res = ''
+  if(m.length>2)return false;	// :: can't be used more than once in IPv6.
+  for (j = 0; j < m.length; j++) {
+    if (m[j].length === 0) {	// Skip if empty.
+      continue
+    }
+    m[j] = m[j].split(':')
+    for (i = 0; i < m[j].length; i++) {
+
```
Comment (kukawski) [inline src/php/network/inet_pton.js:38]: here we could define `let hextet = hextets[i]` to not repeat `m[j][i]` all over the code.

File: src/php/network/inet_pton.js [hunk 2]
```diff
@@ -21,36 +21,44 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
-      for (i = 0; i < m[j].length; i++) {
-        m[j][i] = parseInt(m[j][i], 16)
-        // Would be NaN if it was blank, return false.
-        if (isNaN(m[j][i])) {
-          // Invalid IP.
-          return false
-        }
-        m[j][i] = f(m[j][i] >> 8) + f(m[j][i] & 0xFF)
+
+      m[j][i] = parseInt(m[j][i], 16)
```
Comment (kukawski) [inline src/php/network/inet_pton.js:44]: the result could again be stored in some temporary variable

File: src/php/network/inet_pton.js [hunk 3]
```diff
@@ -21,36 +21,44 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
-    }
-    x = m[1].length + m[3].length
-    if (x === 16) {
-      return m[1] + m[3]
-    } else if (x < 16 && m[2].length > 0) {
-      return m[1] + (new Array(16 - x + 1))
-        .join('\x00') + m[3]
+      m[j][i] = f(m[j][i] >> 8) + f(m[j][i] & 0xFF)
     }
+    m[j] = m[j].join('')
```
Comment (kukawski) [inline src/php/network/inet_pton.js:53]: this assignment is obsolete, you could assign directly `res += m[j].join('')` (or `res += hextets.join('')` as per above suggestions).

**Review by kukawski — CHANGES_REQUESTED** [review]

**Comment by JamieSlome** [comment]
@kukawski - the requested changes have been made!

**Review by kukawski — CHANGES_REQUESTED** [review]
I reviewed your patch again. I have some more comments. Sorry for that.
Also, additionally to the line comments:
- you can add your name in the meta-comments in the top of the file. I believe `improved by:` is a good tag.
- please remove the variable `r`, because it's unused after your changes and linter complains about it.

File: src/php/network/inet_pton.js [hunk 4]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
+  }
+
+  m = a.split('::')
+
+  let res = ''
+  if (m.length > 2) {
+    return false
+  }	// :: can't be used more than once in IPv6.
+
+  let reHexDigits = /^[\da-f]{1,4}$/
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:37]: Done! 👍

File: src/php/network/inet_pton.js [hunk 5]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
+
+  for (j = 0; j < m.length; j++) {
+    if (m[j].length === 0) {	// Skip if empty.
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:40]: Done 👍

File: src/php/network/inet_pton.js [hunk 6]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
+      continue
+    }
+    m[j] = m[j].split(':')
+    for (i = 0; i < m[j].length; i++) {
+      let hextet = m[j][i]
+      // check if valid hex string up to 4 chars
+      if(!reHexDigits.test(hextet)) {
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:47]: Done! 👍

File: src/php/network/inet_pton.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:37]: Done! 👍

File: src/php/network/inet_pton.js [hunk 7]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
+      hextet = f(hextet >> 8, hextet & 0xFF)
     }
+    res += m[j].join('')
+    res += m[j]
+  }
+  x = res.length
+  if (x === 16) {
+    return res
+  } else if (x < 16 && m.length > 1) {
+    return res + '\x00'.repeat(16 - x + 1) + m[1]
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:67]: Done! 👍

File: src/php/network/inet_pton.js [hunk 8]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
       }
-      m[j] = m[j].join('')
-    }
-    x = m[1].length + m[3].length
-    if (x === 16) {
-      return m[1] + m[3]
-    } else if (x < 16 && m[2].length > 0) {
-      return m[1] + (new Array(16 - x + 1))
-        .join('\x00') + m[3]
+      hextet = f(hextet >> 8, hextet & 0xFF)
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:58]: Done! 👍

File: src/php/network/inet_pton.js [hunk 9]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
+  if (a.length > 39) {
+    return false
+  }
+
+  m = a.split('::')
+
+  let res = ''
+  if (m.length > 2) {
+    return false
+  }	// :: can't be used more than once in IPv6.
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:35]: Done! 🍰

File: src/php/network/inet_pton.js [hunk 10]
```diff
@@ -17,40 +17,54 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
-    x = m[1].length + m[3].length
-    if (x === 16) {
-      return m[1] + m[3]
-    } else if (x < 16 && m[2].length > 0) {
-      return m[1] + (new Array(16 - x + 1))
-        .join('\x00') + m[3]
+      hextet = f(hextet >> 8, hextet & 0xFF)
     }
+    res += m[j].join('')
+    res += m[j]
```
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:61]: Done! 👍

File: src/php/network/inet_pton.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (JamieSlome) [reply src/php/network/inet_pton.js:67]: Done! 👍

File: src/php/network/inet_pton.js [hunk 11]
```diff
@@ -55,18 +55,11 @@ module.exports = function inet_pton (a) { // eslint-disable-line camelcase
...
-  x = res.length
-  if (x === 16) {
-    return res
-  } else if (x < 16 && m.length > 1) {
-    return res + '\x00'.repeat(16 - x + 1) + m[1]
-  }
 
-  // Invalid IP
-  return false
+  return m.join('\x00'.repeat(16 - m[0].length - m[1].length))
```
Comment (kukawski) [inline src/php/network/inet_pton.js:64]: `m[1]` can be undefined and reading `.length` of undefined throws error, so let's change that line to

```javascript
  return m.join('\x00'.repeat(16 - m.reduce((tl, m) => tl + m.length, 0)))
```

**Review by kukawski — CHANGES_REQUESTED** [review]
Awesome, we're almost there.
Please remove all occurences or variable `res` and drop variable `x`, as they are not needed anymore + 1 bugfix commented below.

File: src/php/network/inet_pton.js (hunk 11 above)
Comment (JamieSlome) [reply src/php/network/inet_pton.js]: Done! 👍

**Comment by JamieSlome** [comment]
@kukawski - that should be all of the requested changes!

Thanks! 🍰

**Comment by kukawski** [comment]
Thank you once again for the patch.

**Pull request merged**